From a34c9da09b40068b327be9e1b82bc418b3ecc28d Mon Sep 17 00:00:00 2001
From: Miroslav Ondra <ondra@faster.cz>
Date: Tue, 10 Sep 2024 21:57:33 +0200
Subject: [PATCH 2/6] Add support for bootcount in ds1307 rtc chip

---
 drivers/rtc/ds1307.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/rtc/ds1307.c b/drivers/rtc/ds1307.c
index 04921099..76a2a139 100644
--- a/drivers/rtc/ds1307.c
+++ b/drivers/rtc/ds1307.c
@@ -219,6 +219,27 @@ static void rtc_write (uchar reg, uchar val)
 #endif /* !CONFIG_DM_RTC */
 
 #ifdef CONFIG_DM_RTC
+
+static int ds1307_rtc_read8(struct udevice *dev, unsigned int reg)
+{
+	int ret;
+	u8 buf;
+
+	ret = dm_i2c_read(dev, reg, &buf, sizeof(buf));
+
+	if (ret < 0)
+		return ret;
+
+	return buf;
+}
+
+static int ds1307_rtc_write8(struct udevice *dev, unsigned int reg, int val)
+{
+	u8 buf = (u8)val;
+	return dm_i2c_write(dev, reg, &buf, sizeof(buf));
+}
+
+
 static int ds1307_rtc_set(struct udevice *dev, const struct rtc_time *tm)
 {
 	int ret;
@@ -345,6 +366,8 @@ static int ds1307_probe(struct udevice *dev)
 static const struct rtc_ops ds1307_rtc_ops = {
 	.get = ds1307_rtc_get,
 	.set = ds1307_rtc_set,
+	.read8 = ds1307_rtc_read8,
+	.write8 = ds1307_rtc_write8,
 	.reset = ds1307_rtc_reset,
 };
 
-- 
2.39.5

