From ec230504602be98a4a5e5a83ca346bd46068e7ab Mon Sep 17 00:00:00 2001
From: Miroslav Ondra <ondra@faster.cz>
Date: Fri, 27 Sep 2024 02:04:14 +0200
Subject: [PATCH] Fix amba-pl011 serial driver - remove udelay in RS485

---
 drivers/tty/serial/amba-pl011.c | 201 ++++++++++++++++++++++----------
 1 file changed, 141 insertions(+), 60 deletions(-)

diff --git a/drivers/tty/serial/amba-pl011.c b/drivers/tty/serial/amba-pl011.c
index 8e91d74f2..c36106b0c 100644
--- a/drivers/tty/serial/amba-pl011.c
+++ b/drivers/tty/serial/amba-pl011.c
@@ -264,6 +264,13 @@ struct pl011_dmatx_data {
 	bool			queued;
 };
 
+enum pl011_tx_state {
+	OFF,
+	WAIT_AFTER_RTS,
+	SEND,
+	WAIT_AFTER_SEND,
+};
+
 /*
  * We wrap our port structure around the generic uart_port.
  */
@@ -278,8 +285,10 @@ struct uart_amba_port {
 	unsigned int		fifosize;	/* vendor-specific */
 	unsigned int		fixed_baud;	/* vendor-set fixed baud rate */
 	char			type[12];
-	bool			rs485_tx_started;
-	unsigned int		rs485_tx_drain_interval; /* usecs */
+	ktime_t			rs485_tx_drain_interval; /* nano */
+	enum pl011_tx_state	tx_state;
+	struct hrtimer		trigger_start_tx;
+	struct hrtimer		trigger_stop_tx;
 #ifdef CONFIG_DMA_ENGINE
 	/* DMA stuff */
 	bool			using_tx_dma;
@@ -290,6 +299,8 @@ struct uart_amba_port {
 #endif
 	void *led_trigger_data;
 	struct mctrl_gpios *gpios;
+
+
 };
 
 static unsigned int pl011_tx_empty(struct uart_port *port);
@@ -1304,45 +1315,45 @@ static inline bool pl011_dma_rx_running(struct uart_amba_port *uap)
 
 static void pl011_rs485_tx_stop(struct uart_amba_port *uap)
 {
-	/*
-	 * To be on the safe side only time out after twice as many iterations
-	 * as fifo size.
-	 */
-	const int MAX_TX_DRAIN_ITERS = uap->port.fifosize * 2;
 	struct uart_port *port = &uap->port;
-	int i = 0;
 	u32 cr;
 
-	/* Wait until hardware tx queue is empty */
-	while (!pl011_tx_empty(port)) {
-		if (i > MAX_TX_DRAIN_ITERS) {
-			dev_warn(port->dev,
-				 "timeout while draining hardware tx queue\n");
-			break;
-		}
+	if ((uap->tx_state == SEND) || (uap->tx_state == WAIT_AFTER_SEND)) {
+		uap->tx_state = WAIT_AFTER_SEND;
 
-		udelay(uap->rs485_tx_drain_interval);
-		i++;
+		/* Wait until hardware tx queue is empty */
+		if (!pl011_tx_empty(port)) {
+			hrtimer_start(&uap->trigger_stop_tx, uap->rs485_tx_drain_interval,
+			              HRTIMER_MODE_REL);
+			return;
+		}
+		if (port->rs485.delay_rts_after_send > 0) {
+			hrtimer_start(&uap->trigger_stop_tx, ms_to_ktime(port->rs485.delay_rts_after_send),
+			              HRTIMER_MODE_REL);
+			return;
+		}
+		/* continue without any delay */
 	}
 
-	if (port->rs485.delay_rts_after_send)
-		mdelay(port->rs485.delay_rts_after_send);
+	if (uap->tx_state == WAIT_AFTER_RTS || uap->tx_state == WAIT_AFTER_SEND) {
 
-	cr = pl011_read(uap, REG_CR);
+		hrtimer_try_to_cancel(&uap->trigger_start_tx);
 
-	if (port->rs485.flags & SER_RS485_RTS_AFTER_SEND) {
-		mctrl_gpio_set(uap->gpios, uap->port.mctrl | TIOCM_RTS);
-		cr &= ~UART011_CR_RTS;
-	} else {
-		mctrl_gpio_set(uap->gpios, uap->port.mctrl & ~TIOCM_RTS);
-		cr |= UART011_CR_RTS;
-	}
-	/* Disable the transmitter and reenable the transceiver */
-	cr &= ~UART011_CR_TXE;
-	cr |= UART011_CR_RXE;
-	pl011_write(cr, uap, REG_CR);
+		cr = pl011_read(uap, REG_CR);
 
-	uap->rs485_tx_started = false;
+		if (port->rs485.flags & SER_RS485_RTS_AFTER_SEND) {
+			mctrl_gpio_set(uap->gpios, uap->port.mctrl | TIOCM_RTS);
+			cr &= ~UART011_CR_RTS;
+		} else {
+			mctrl_gpio_set(uap->gpios, uap->port.mctrl & ~TIOCM_RTS);
+			cr |= UART011_CR_RTS;
+		}
+		/* Disable the transmitter and reenable the transceiver */
+		cr &= ~UART011_CR_TXE;
+		cr |= UART011_CR_RXE;
+		pl011_write(cr, uap, REG_CR);
+		uap->tx_state = OFF;
+	}
 }
 
 static void pl011_stop_tx(struct uart_port *port)
@@ -1350,12 +1361,20 @@ static void pl011_stop_tx(struct uart_port *port)
 	struct uart_amba_port *uap =
 	    container_of(port, struct uart_amba_port, port);
 
-	uap->im &= ~UART011_TXIM;
-	pl011_write(uap->im, uap, REG_IMSC);
-	pl011_dma_tx_stop(uap);
+	if (uap->tx_state == OFF)
+		return;
 
-	if ((port->rs485.flags & SER_RS485_ENABLED) && uap->rs485_tx_started)
+	if (uap->tx_state == SEND) {
+		uap->im &= ~UART011_TXIM;
+		pl011_write(uap->im, uap, REG_IMSC);
+		pl011_dma_tx_stop(uap);
+	}
+
+	if (port->rs485.flags & SER_RS485_ENABLED)  {
 		pl011_rs485_tx_stop(uap);
+	} else {
+		uap->tx_state = OFF;
+	}
 }
 
 static bool pl011_tx_chars(struct uart_amba_port *uap, bool from_irq);
@@ -1369,32 +1388,51 @@ static void pl011_start_tx_pio(struct uart_amba_port *uap)
 	}
 }
 
-static void pl011_rs485_tx_start(struct uart_amba_port *uap)
+static int pl011_rs485_tx_start(struct uart_amba_port *uap)
 {
 	struct uart_port *port = &uap->port;
 	u32 cr;
 
-	/* Enable transmitter */
-	cr = pl011_read(uap, REG_CR);
-	cr |= UART011_CR_TXE;
+	if (uap->tx_state == OFF) {
 
-	/* Disable receiver if half-duplex */
-	if (!(port->rs485.flags & SER_RS485_RX_DURING_TX))
-		cr &= ~UART011_CR_RXE;
+		cr = pl011_read(uap, REG_CR);
 
-	if (port->rs485.flags & SER_RS485_RTS_ON_SEND) {
-		mctrl_gpio_set(uap->gpios, uap->port.mctrl | TIOCM_RTS);
-		cr &= ~UART011_CR_RTS;
-	} else {
-		mctrl_gpio_set(uap->gpios, uap->port.mctrl & ~TIOCM_RTS);
-		cr |= UART011_CR_RTS;
+		/* Disable receiver if half-duplex */
+		if (!(port->rs485.flags & SER_RS485_RX_DURING_TX))
+			cr &= ~UART011_CR_RXE;
+
+		if (port->rs485.flags & SER_RS485_RTS_ON_SEND) {
+			mctrl_gpio_set(uap->gpios, uap->port.mctrl | TIOCM_RTS);
+			cr &= ~UART011_CR_RTS;
+		} else {
+			mctrl_gpio_set(uap->gpios, uap->port.mctrl & ~TIOCM_RTS);
+			cr |= UART011_CR_RTS;
+		}
+		pl011_write(cr, uap, REG_CR);
+
+		uap->tx_state = WAIT_AFTER_RTS;
+
+		if (port->rs485.delay_rts_before_send > 0) {
+			hrtimer_start(&uap->trigger_start_tx,
+			              ms_to_ktime(port->rs485.delay_rts_before_send),
+			              HRTIMER_MODE_REL);
+			return 1;
+		}
+		/* continue without any delay */
 	}
-	pl011_write(cr, uap, REG_CR);
+	if (uap->tx_state == WAIT_AFTER_SEND
+	    || uap->tx_state == WAIT_AFTER_RTS) {
 
-	if (port->rs485.delay_rts_before_send)
-		mdelay(port->rs485.delay_rts_before_send);
+		hrtimer_try_to_cancel(&uap->trigger_stop_tx);
 
-	uap->rs485_tx_started = true;
+		/* Enable transmitter */
+		cr = pl011_read(uap, REG_CR);
+		cr |= UART011_CR_TXE;
+		pl011_write(cr, uap, REG_CR);
+	}
+	uap->tx_state = SEND;
+
+	return 0;
 }
 
 static void pl011_start_tx(struct uart_port *port)
@@ -1402,14 +1440,45 @@ static void pl011_start_tx(struct uart_port *port)
 	struct uart_amba_port *uap =
 	    container_of(port, struct uart_amba_port, port);
 
-	if ((uap->port.rs485.flags & SER_RS485_ENABLED) &&
-	    !uap->rs485_tx_started)
-		pl011_rs485_tx_start(uap);
+	if (uap->port.rs485.flags & SER_RS485_ENABLED) {
+		if (pl011_rs485_tx_start(uap)!=0)
+			return;
+	} else
+		uap->tx_state = SEND;
 
 	if (!pl011_dma_tx_start(uap))
 		pl011_start_tx_pio(uap);
 }
 
+static enum hrtimer_restart pl011_trigger_start_tx(struct hrtimer *t)
+{
+	struct uart_amba_port *uap =
+	    container_of(t, struct uart_amba_port, trigger_start_tx);
+	unsigned long flags;
+
+	spin_lock_irqsave(&uap->port.lock, flags);
+	if (uap->tx_state == WAIT_AFTER_RTS)
+		pl011_start_tx(&uap->port);
+	spin_unlock_irqrestore(&uap->port.lock, flags);
+
+	return HRTIMER_NORESTART;
+}
+
+static enum hrtimer_restart pl011_trigger_stop_tx(struct hrtimer *t)
+{
+	struct uart_amba_port *uap =
+	    container_of(t, struct uart_amba_port, trigger_stop_tx);
+	unsigned long flags;
+
+	spin_lock_irqsave(&uap->port.lock, flags);
+	if (uap->tx_state == WAIT_AFTER_SEND)
+		pl011_stop_tx(&uap->port);
+	spin_unlock_irqrestore(&uap->port.lock, flags);
+
+	return HRTIMER_NORESTART;
+}
+
+
 static void pl011_stop_rx(struct uart_port *port)
 {
 	struct uart_amba_port *uap =
@@ -1448,7 +1517,6 @@ __acquires(&uap->port.lock)
 
 	spin_unlock(&uap->port.lock);
 	tty_flip_buffer_push(&uap->port.state->port);
-	//rxtx_rx_led_activity(uap->led_trigger_data, 1);
 
 	/*
 	 * If we were temporarily out of DMA mode for a while,
@@ -1900,7 +1968,6 @@ static int pl011_startup(struct uart_port *port)
 
 	if (!(port->rs485.flags & SER_RS485_ENABLED))
 		cr |= UART011_CR_TXE;
-
 	pl011_write(cr, uap, REG_CR);
 
 	spin_unlock_irq(&uap->port.lock);
@@ -2000,7 +2067,7 @@ static void pl011_shutdown(struct uart_port *port)
 
 	pl011_dma_shutdown(uap);
 
-	if ((port->rs485.flags & SER_RS485_ENABLED) && uap->rs485_tx_started)
+	if ((port->rs485.flags & SER_RS485_ENABLED))
 		pl011_rs485_tx_stop(uap);
 
 	free_irq(uap->port.irq, uap);
@@ -2145,7 +2212,7 @@ pl011_set_termios(struct uart_port *port, struct ktermios *termios,
 	 * with the given baud rate. We use this as the poll interval when we
 	 * wait for the tx queue to empty.
 	 */
-	uap->rs485_tx_drain_interval = DIV_ROUND_UP(bits * 1000 * 1000, baud);
+	uap->rs485_tx_drain_interval = ns_to_ktime(DIV_ROUND_UP(bits * NSEC_PER_SEC, baud));
 
 	pl011_setup_status_masks(port, termios);
 
@@ -2278,8 +2345,17 @@ static int pl011_rs485_config(struct uart_port *port, struct ktermios *termios,
 		u32 cr = pl011_read(uap, REG_CR);
 
 		cr &= ~UART011_CR_RTSEN;
+
+		if (rs485->flags & SER_RS485_RTS_AFTER_SEND) {
+			mctrl_gpio_set(uap->gpios, uap->port.mctrl | TIOCM_RTS);
+			cr &= ~UART011_CR_RTS;
+		} else if (rs485->flags & SER_RS485_RTS_ON_SEND) {
+			mctrl_gpio_set(uap->gpios, uap->port.mctrl & ~TIOCM_RTS);
+			cr |= UART011_CR_RTS;
+		}
 		pl011_write(cr, uap, REG_CR);
 		port->status &= ~UPSTAT_AUTORTS;
+
 	}
 
 	return 0;
@@ -2877,6 +2953,11 @@ static int pl011_probe(struct amba_device *dev, const struct amba_id *id)
 		}
 	}
 
+	hrtimer_init(&uap->trigger_start_tx, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+	hrtimer_init(&uap->trigger_stop_tx, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+	uap->trigger_start_tx.function = pl011_trigger_start_tx;
+	uap->trigger_stop_tx.function = pl011_trigger_stop_tx;
+
 	ret = pl011_setup_port(&dev->dev, uap, &dev->res, portnr);
 	if (ret)
 		return ret;
-- 
2.39.5

